From c205a42958c993b230200199209edb745a3fe420 Mon Sep 17 00:00:00 2001
From: Mykola Ostrovskyy <mykola@ti.com>
Date: Tue, 8 Apr 2014 14:15:42 -0500
Subject: [PATCH 08/11] HWComposer: Prevent FB de-allocation with HWC 1.1+

Even though HWComposer will not use FB HAL functionality when
working with newer HWC API, the FB instance should be keept alive
to prevent destruction of flip chain.

Change-Id: I606873fc5d03f676d3a256269e8503d5de5df43e
Signed-off-by: Mykola Ostrovskyy <mykola@ti.com>
---
 services/surfaceflinger/DisplayHardware/HWComposer.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/services/surfaceflinger/DisplayHardware/HWComposer.cpp b/services/surfaceflinger/DisplayHardware/HWComposer.cpp
index 0859149..d67dab3 100644
--- a/services/surfaceflinger/DisplayHardware/HWComposer.cpp
+++ b/services/surfaceflinger/DisplayHardware/HWComposer.cpp
@@ -109,6 +109,10 @@ HWComposer::HWComposer(
     int fberr = loadFbHalModule();
     loadHwcModule();
 
+#ifdef OMAP_ENHANCEMENT
+    // FB HAL must stay open independent of HWC API version. Closing FB HAL will
+    // result in destruction of flip chain and de-allocation of framebuffer.
+#else
     if (mFbDev && mHwc && hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1)) {
         // close FB HAL if we don't needed it.
         // FIXME: this is temporary until we're not forced to open FB HAL
@@ -116,6 +120,7 @@ HWComposer::HWComposer(
         framebuffer_close(mFbDev);
         mFbDev = NULL;
     }
+#endif
 
     // If we have no HWC, or a pre-1.1 HWC, an FB dev is mandatory.
     if ((!mHwc || !hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1))
@@ -164,7 +169,11 @@ HWComposer::HWComposer(
         }
     }
 
+#ifdef OMAP_ENHANCEMENT
+    if (!mHwc || !hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1)) {
+#else
     if (mFbDev) {
+#endif
         ALOG_ASSERT(!(mHwc && hwcHasApiVersion(mHwc, HWC_DEVICE_API_VERSION_1_1)),
                 "should only have fbdev if no hwc or hwc is 1.0");
 
-- 
2.7.4

